clients/src/main/java/org/apache/kafka/common/telemetry/internals/ClientTelemetryReporter.java

```diff
@@ -0,0 +1,976 @@
+    }
+
+    private void initCollectors() {
+        kafkaMetricsCollector = new KafkaMetricsCollector(
+            TelemetryMetricNamingConvention.getClientTelemetryMetricNamingStrategy(
+                telemetryProvider.domain()));
+    }
+
+    private ResourceMetrics buildMetric(Metric metric) {
+        return ResourceMetrics.newBuilder()
+            .setResource(telemetryProvider.resource())
+            .addScopeMetrics(ScopeMetrics.newBuilder()
+                .addMetrics(metric)
+                .build()).build();
+    }
+
+    // Visible for testing, only for unit tests
+    MetricsCollector metricsCollector() {
+        return kafkaMetricsCollector;
+    }
+
+    // Visible for testing, only for unit tests
+    ClientTelemetryProvider telemetryProvider() {
+        return telemetryProvider;
+    }
+
+    class DefaultClientTelemetrySender implements ClientTelemetrySender {
+
+        /*
+         These are the lower and upper bounds of the jitter that we apply to the initial push
+         telemetry API call. This helps to avoid a flood of requests all coming at the same time.
+        */
+        private final static double INITIAL_PUSH_JITTER_LOWER = 0.5;
+        private final static double INITIAL_PUSH_JITTER_UPPER = 1.5;
+
+        private final ReadWriteLock lock = new ReentrantReadWriteLock();
+        private final Condition subscriptionLoaded = lock.writeLock().newCondition();
+        private final Condition terminalPushInProgress = lock.writeLock().newCondition();
+        /*
+         Initial state should be subscription needed which should allow issuing first telemetry
+         request of get telemetry subscription.
+        */
+        private ClientTelemetryState state = ClientTelemetryState.SUBSCRIPTION_NEEDED;
+
+        private ClientTelemetrySubscription subscription;
+
+        /*
+         Last time a telemetry request was made. Initialized to 0 to indicate that no request has
+         been made yet. Telemetry requests, get or post, should always be made after the push interval
+         time has elapsed.
+        */
+        private long lastRequestMs;
+        /*
+         Interval between telemetry requests in milliseconds. Initialized to 0 to indicate that the
+         interval has not yet been computed. The first get request will be immediately triggered as
+         soon as the client is ready.
+        */
+        private int intervalMs;
+        /*
+         Whether the client telemetry sender is enabled or not. Initialized to true to indicate that
+         the client telemetry sender is enabled. This is used to disable the client telemetry sender
+         when the client receives unrecoverable error from broker.
+        */
+        private boolean enabled;
+
+        private DefaultClientTelemetrySender() {
+            enabled = true;
+        }
+
+        @Override
+        public long timeToNextUpdate(long requestTimeoutMs) {
+            final long now = time.milliseconds();
```

Comment:
with unit is better - nowMs

Reply:
Though mostly clients code uses `now` for `time.milliseconds()` hence followed the same, there are some occurrences on `nowMs` as well. I do not have any strong opinion on either hence changed it to `nowMs`.